### PR TITLE
Test on windows and mac on each PR and add openfermion to dev requirements

### PR DESCRIPTION
**Pull request by andreamari**
Description
-----------
Often PRs pass all tests but they fail when merged into master since Windows and Mac are tested only on master.
The original reason for non testing Mac and Windows in each PR was to save time.
It may be worth to reconsider this decision depending on the testing time and considering that the main bottleneck at the moment is the docs build and not pytest.

If some items remain, you can mark this a [draft pull request](https://github.blog/2019-02-14-introducing-draft-pull-requests/).

License
-------

- [x] I license this contribution under the terms of the GNU GPL, version 3 and grant Unitary Fund the right to provide additional permissions as described in section 7 of the GNU GPL, version 3.

Tips
----

- If the validation check fails:

    1. Run `make check-types` (from the root directory of the repository) and fix any [mypy](https://mypy.readthedocs.io/en/stable/) errors.

    2. Run `make check-style` and fix any [flake8](http://flake8.pycqa.org) errors.

    3. Run `make format` to format your code with the [black](https://black.readthedocs.io/en/stable/index.html) autoformatter.

  For more information, check the [Mitiq style guidelines](https://mitiq.readthedocs.io/en/stable/contributing.html#style-guidelines).

- Write "Fixes #XYZ" in the description if this PR fixes Issue #XYZ.

### CODE REVIEW

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -138,3 +138,24 @@ jobs:
         uses: codecov/codecov-action@v3.1.0
         with:
           fail_ci_if_error: true
+
+  test-nonlinux:
+    runs-on: ${{ matrix.os }}
+    strategy:
+      matrix:
+        os: [macos-latest, windows-latest]
+        python-version: [3.7, 3.8]
```
Comment (natestemen) [inline .github/workflows/test.yml:147]: Minor nitpick: I think this is slightly more clear if the `runs-on` attribute comes _after_ the strategy matrix as is done in the [github documentation](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#example-using-a-multi-dimension-matrix) on this.

**Review by natestemen — APPROVED** [review]
Looks good, just one minor nitpick that's not essential.

The main question I had is whether the new `test-nonlinux` step can be merged with `test-linux` so that the testing is the same regardless of OS. We would still only want to submit the code coverage report once, but that should work just fine I think. Is there a reason to keep these separate?

**Comment by andreamari** [comment]
> Looks good, just one minor nitpick that's not essential.
> 
> The main question I had is whether the new `test-nonlinux` step can be merged with `test-linux` so that the testing is the same regardless of OS. We would still only want to submit the code coverage report once, but that should work just fine I think. Is there a reason to keep these separate?

The reason I kept them separated is that there are several differences in the types of instructions (`make test` vs `make test-all`, using Docker vs not using it, codecove). The original idea was to test _everything_ on linux and to run a lighter test suite on non-linux platforms. 

It is worth exploring the possibility of running exactly the same types of tests on linux ad non-linux. But it is probably a topic for a new issue/PR.

**Comment by natestemen** [comment]
Makes sense, thanks for the explanation. :shipit:

**Pull request merged**